pacsman/pynetdicom_client.py

```diff
@@ -0,0 +1,399 @@
+                return result_path if os.path.exists(result_path) else None
+            except Exception as e:
+                raise e
+            finally:
+                scp.stop()
+
+
+def _call_c_find_patients(assoc, search_field, search_query):
+    dataset = Dataset()
+
+    dataset.PatientID = None
+    dataset.PatientName = ''
+    dataset.PatientBirthDate = None
+    dataset.StudyDate = ''
+    dataset.StudyInstanceUID = ''
+    dataset.QueryRetrieveLevel = 'STUDY'
+
+    setattr(dataset, search_field, search_query)
+
+    return assoc.send_c_find(dataset, query_model='S')
+
+
+class StorageSCP(threading.Thread):
+    def __init__(self, client_ae, result_dir):
+        self.result_dir = result_dir
+
+        self.ae_title = f'{client_ae}-SCP'
+        self.ae = AE(ae_title=self.ae_title,
+                     port=11113,
+                     transfer_syntax=[ExplicitVRLittleEndian],
+                     scp_sop_class=[x for x in StorageSOPClassList])
+
+        self.ae.on_c_store = self._on_c_store
+
+        threading.Thread.__init__(self)
+
+        self.daemon = True
+
+    def run(self):
+        """The thread run method"""
+        self.ae.start()
+
+    def stop(self):
+        """Stop the SCP thread"""
+        self.ae.stop()
+
+    def _on_c_store(self, dataset, context, info):
+        '''
+        :param dataset: pydicom.Dataset
+            The DICOM dataset sent via the C-STORE
+        :param context: pynetdicom3.presentation.PresentationContextTuple
+            Details of the presentation context the dataset was sent under.
+        :param info: dict
+            A dict containing information about the association and DIMSE message.
+        :return: pynetdicom.sop_class.Status or int
+        '''
+        try:
+
+            os.makedirs(self.result_dir, exist_ok=True)
+
+            filename = f'{dataset.SOPInstanceUID}.dcm'
+            filepath = os.path.join(self.result_dir, filename)
+
+            logger.info(f'Storing DICOM file: {filepath}')
+
+            if os.path.exists(filename):
+                logger.warning('DICOM file already exists, overwriting')
+
+            meta = Dataset()
+            meta.MediaStorageSOPClassUID = dataset.SOPClassUID
+            meta.MediaStorageSOPInstanceUID = dataset.SOPInstanceUID
+            meta.ImplementationClassUID = pynetdicom_implementation_uid
+            meta.TransferSyntaxUID = context.transfer_syntax
+
+            # The following is not mandatory, set for convenience
+            meta.ImplementationVersionName = pynetdicom_version
+
+            ds = FileDataset(filepath, {}, file_meta=meta, preamble=b"\0" * 128)
+            ds.update(dataset)
+            ds.is_little_endian = context.transfer_syntax.is_little_endian
+
+            ds.is_implicit_VR = context.transfer_syntax.is_implicit_VR
+            ds.save_as(filepath, write_like_original=False)
+
+            status_ds = Dataset()
+            status_ds.Status = 0x0000
+            return status_ds
```

Comment:
Perhaps we can combine the return statements by pulling them out of the try/except?

Reply:
Completed in latest.